### PR TITLE
Switch

### PR DESCRIPTION
**Pull request by emdrayallday**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/components/Input/components/Switch/index.tsx [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import React, { ReactElement } from 'react';
+import clsx from 'clsx';
+import InputBase from '../InputBase';
+import { Props } from './types';
+
+
+function Switch({ className, ...otherProps }: Props): ReactElement {
+    const switchClasses = clsx(
+        'brew-Input--Switch',
```
Comment (lawren) [inline src/components/Input/components/Switch/index.tsx]: `--` denotes a state or condition. In this case, I think it's safe to just use the class `brew-Switch` since we likely won't have any other component using that name.

File: src/components/Input/components/Switch/index.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        className,
+    );
+    const customControl = (
+        <div className="brew-Input__customControl">
```
Comment (lawren) [inline src/components/Input/components/Switch/index.tsx]: This class and the ones below should start with `brew-Switch__` and I would avoid using any reference to `customControl` as that is a reference to the inner working of the component. 

`brew-Switch__control`, `brew-Switch__track`, and `brew-Switch__knob` should suffice.

File: src/components/Input/components/Switch/index.tsx [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import React, { ReactElement } from 'react';
+import clsx from 'clsx';
+import InputBase from '../InputBase';
+import { Props } from './types';
+
+
+function Switch({ className, ...otherProps }: Props): ReactElement {
```
Comment (lawren) [inline src/components/Input/components/Switch/index.tsx]: For consistency's sake, the `props` should always be multiline if there is more than one and the block should be padded with two linebreaks.

File: src/components/Input/components/Switch/index.tsx [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    const switchClasses = clsx(
+        'brew-Input--Switch',
+        className,
+    );
+    const customControl = (
```
Comment (lawren) [inline src/components/Input/components/Switch/index.tsx]: Since these two constants are unrelated to each other, I would advise adding padding between them for readability purposes.

File: src/components/Input/components/Switch/stories.tsx [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import React, { ReactElement } from 'react';
+import Switch from './index';
+import { boolean, text } from '@storybook/addon-knobs';
+
```
Comment (lawren) [inline src/components/Input/components/Switch/stories.tsx]: There should be double padding after imports.

File: src/components/Input/components/Switch/stories.tsx [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+export default {
+    component: Switch,
+    title: 'Components|User Input/Input/Switch',
+};
+
```
Comment (lawren) [inline src/components/Input/components/Switch/stories.tsx]: and double padding between these two blocks as well.

File: src/components/Input/components/Switch/types.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import { Props as InputBaseProps } from '../InputBase/types';
+
+
+export type Props = Omit<InputBaseProps,
+    | 'before'
```
Comment (joshrcook) [inline src/components/Input/components/Switch/types.ts]: `after` shouldn't be an allowed prop

File: src/components/Input/components/Switch/types.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+    | 'children'
+    | 'cols'
+    | 'type'
+    | 'as'
+    | 'defaultValue'
+    | 'placeholder'
+    | 'rows'
+    | 'value'
```
Comment (joshrcook) [inline src/components/Input/components/Switch/types.ts]: `value` should be an allowed prop

File: src/components/Input/components/Switch/index.tsx [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+
+function Switch({
+    className,
+    ...otherProps
+}: Props): ReactElement {
+
+
+    const switchClasses = clsx(
+        'brew-Switch',
```
Comment (joshrcook) [inline src/components/Input/components/Switch/index.tsx]: The base classname should be `brew-Input--Switch` since it's an input component

File: src/components/Input/components/Switch/index.tsx [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+        className,
+    );
+
+
+    const customControl = (
+        <div className="brew-Switch__custom">
```
Comment (joshrcook) [inline src/components/Input/components/Switch/index.tsx]: These classes should be something like `brew-Input__custom`, etc.

File: theme/cerveza/src/components/_Input.scss [hunk 11]
```diff
@@ -363,3 +363,106 @@
...
+
+$switchSize: size(xx-large);
+
+.brew-Switch {
+    display: inline-flex;
+    height: $switchSize;
+    position: relative;
+    cursor: pointer;
+
+    div {
```
Comment (joshrcook) [inline theme/cerveza/src/components/_Input.scss]: We wouldn't want to have CSS that just references an HTML tag.  Everything should be class based so that the HTML structure can change while the styles stay the same.

File: theme/cerveza/src/components/_Input.scss [hunk 12]
```diff
@@ -363,3 +363,106 @@
...
+        width: $switchSize * 2;
+    }
+
+    input {
+        position: absolute;
+        opacity: 0;
+        height: 0;
+        width: 0;
+
+        &:checked ~ div .brew-Switch__knob {
```
Comment (joshrcook) [inline theme/cerveza/src/components/_Input.scss]: Same thing here.  General `div` selectors aren't good practice

File: src/components/Input/components/Switch/index.tsx (hunk 9 above)
Comment (lawren) [reply src/components/Input/components/Switch/index.tsx]: Based on BEM though, that is saying that `Switch` is a modifier of `Input`, which it's really not considering it's a truly separate component. We also won't have another `Switch` component, so unless we want to take the approach that we use parent categories in our BEM bases throughout, I think it's fine as is. 

If we do change it, I would advocated `brew-Input-Switch` instead of the modifier syntax.
